File: tests/cases/ruby/report.rb
class Report
  def match(users, orders)
    users.each do |u|
      orders.each do |o| # expect: nested-loop
        puts o if o.user_id == u.id
      end
    end
  end

  def walk(orders)
    orders.each do |order|
      order.lines.each { |line| puts line }
    end
  end

  def selected(items, ids)
    items.select { |i| ids.include?(i.id) } # expect: membership-in-loop
  end

  def load(ids)
    ids.each do |id|
      User.find_by(id: id) # expect: io-or-query-in-loop
    end
  end

  def by_id(items)
    items.inject({}) { |h, item| h.merge(item.id => item) } # expect: quadratic-accumulation
  end

  def by_id_fast(items)
    items.each_with_object({}) { |item, h| h[item.id] = item }
  end

  def preload(ids)
    User.where(id: ids).to_a
  end
end

File: tests/cases/java/Report.java
import java.util.List;
import java.util.Set;
import java.util.regex.Pattern;

class Report {
    String render(List<Row> rows) {
        String out = "";
        for (Row row : rows) {
            out += row.name(); // expect: string-concat-in-loop
        }
        return out;
    }

    String renderFast(List<Row> rows) {
        StringBuilder sb = new StringBuilder();
        for (Row row : rows) {
            sb.append(row.name());
        }
        return sb.toString();
    }

    void match(List<User> users, List<Order> orders) {
        for (User u : users) {
            for (Order o : orders) { // expect: nested-loop
                System.out.println(o);
            }
        }
    }

    void listContains(List<String> names, List<String> banned) {
        for (String n : names) {
            if (banned.contains(n)) { // expect: membership-in-loop
                System.out.println(n);
            }
        }
    }

    void setContains(List<String> names, Set<String> banned) {
        for (String n : names) {
            if (banned.contains(n)) {
                System.out.println(n);
            }
        }
    }

    void pattern(List<String> lines) {
        for (String line : lines) {
            Pattern p = Pattern.compile("\\d+"); // expect: regex-compile-in-loop
            p.matcher(line);
        }
    }

    void drain(List<Job> jobs) {
        while (!jobs.isEmpty()) {
            Job job = jobs.remove(0); // expect: list-shift-in-loop
        }
    }
}
